### PR TITLE
Add "create new service" action

### LINKED ISSUE #5: Create service
As I user, I should be able to create a service.

Should probably lead to a default YAML payload to be edited by user

### PR DESCRIPTION
**Pull request by lstocchi**
it resolves #5 

it has double tabs, one basic and one more advanced with the editor.
The editor has code completion using the service schema taken from vscode.
The dialog has an event log panel that pop ups if there are errors during saving.
The editor content is updated if the fields in the `basic` tab are updated but it's not true the opposite. Not sure if we want to implement it as well but i expect that if a user writes YAML code he doesn't use the basic fields.
Below an example

create-new-service

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/knative/utils/DeployModel.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public JsonNode getSpec() {
+        return spec;
+    }
+
+    public CustomResourceDefinitionContext getCrdContext() {
+        return crdContext;
+    }
+}
```
Comment (evidolob) [inline src/main/java/com/redhat/devtools/intellij/knative/utils/DeployModel.java:48]: Could you add new line?

File: src/main/java/com/redhat/devtools/intellij/knative/validation/KubernetesTypeInfo.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    public static KubernetesTypeInfo fromFileName(String filename) {
+        int index = filename.indexOf('_');
+        String apiGroup = (index != (-1))?filename.substring(0, index):"";
+        String kind = (index != (-1))?filename.substring(index + 1):filename;
+        index = kind.lastIndexOf('.');
+        kind = (index != (-1))?kind.substring(0, index):kind;
+        return new KubernetesTypeInfo(apiGroup, kind);
+
+    }
+}
```
Comment (evidolob) [inline src/main/java/com/redhat/devtools/intellij/knative/validation/KubernetesTypeInfo.java:131]: Could you add new line?

File: src/main/resources/META-INF/plugin-json.xml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<idea-plugin>
+    <extensions defaultExtensionNs="JavaScript.JsonSchema">
+        <ProviderFactory implementation="com.redhat.devtools.intellij.knative.validation.KnSchemasProviderFactory"/>
+    </extensions>
+</idea-plugin>
```
Comment (evidolob) [inline src/main/resources/META-INF/plugin-json.xml:5]: Could you add new line?

File: src/main/resources/schemas/index.properties [hunk 4]
```diff
@@ -0,0 +1 @@
+service.json
```
Comment (evidolob) [inline src/main/resources/schemas/index.properties:1]: same there

File: src/main/resources/snippets/service.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: "add service name"
+spec:
+  template:
+    spec:
+      containers:
+        - image: "add image url" # Reference to the image of the app
+          env:
+            - name: TARGET # The environment variable printed out by the sample app
+              value: "Sample"
```
Comment (evidolob) [inline src/main/resources/snippets/service.yaml:12]: and there

File: src/main/resources/snippets/service.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: serving.knative.dev/v1 # Current version of Knative
+kind: Service
+metadata:
+  name: "add service name"
```
Comment (evidolob) [inline src/main/resources/snippets/service.yaml:4]: I think we should have `namespace` property there, as Schema validator complains about it:
Screenshot 2021-03-17 at 15 10 19

File: src/main/java/com/redhat/devtools/intellij/knative/actions/KnAction.java [hunk 7]
```diff
@@ -41,6 +41,6 @@ private Kn getKn(AnActionEvent anActionEvent) throws IOException {
         return ((KnRootNode)((KnTreeStructure)tree.getClientProperty(Constants.STRUCTURE_PROPERTY)).getRootElement()).getKn();
     }
 
-    public void actionPerformed(AnActionEvent anActionEvent, TreePath path, Object selected, Kn tkn) {}
+    public void actionPerformed(AnActionEvent anActionEvent, TreePath path, Object selected, Kn kn) {}
 
 }
```
Comment (evidolob) [inline src/main/java/com/redhat/devtools/intellij/knative/actions/KnAction.java:46]: Could you add new line for this file?

**Review by evidolob — COMMENTED** [review]
Jus few minor things:

Could we have some sort of validation of `Name` and `Image` fields. 
When I try to create service with name`SomeName`  I got error:
```
Failure executing: POST at: https://api.crc.testing:6443/apis/serving.knative.dev/v1/namespaces/default/services. Message: Service.serving.knative.dev "SomeName" is invalid: metadata.name: Invalid value: "SomeName": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'). Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=metadata.name, message=Invalid value: "SomeName": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'),
```
I think we can use that regexp to validate name before we attempt to crate service

**Comment by lstocchi** [comment]
> Could we have some sort of validation of `Name` and `Image` fields.
> When I try to create service with name`SomeName` I got error:
> 
> ```
> Failure executing: POST at: https://api.crc.testing:6443/apis/serving.knative.dev/v1/namespaces/default/services. Message: Service.serving.knative.dev "SomeName" is invalid: metadata.name: Invalid value: "SomeName": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'). Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=metadata.name, message=Invalid value: "SomeName": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'),
> ```
> 
> I think we can use that regexp to validate name before we attempt to create service.
>
> 
> Which is strange that form allow me to click `Add` button without filling any inputs.

We did something similar in IJ Tekton. The reasoning behind it is that these validations are made by the server (not the front-end). Let's say we add a naming regex rule in the UI, can we guarantee it will always be up-to-date with every cluster? Maybe the current k8s version has a naming rule and the still-not-released version of k8s will have a new one. We decided to just show the error on the ui to make our life easier and let the server do all the work. I get your point as i had the same thought when working on it in IJ tekton. Maybe we can customize the error message if it's too scary? 
@jeffmaury What do you think?

**Comment by evidolob** [comment]
Maybe we can include validation regexp in to JSON Schema for resource?

**Review by jeffmaury — COMMENTED** [review]
There are several questions opened to me by this PR:
- seems there is some code duplicated from IJ Tekton so this is a target to IJ Common
- how the schema(s) has/have been generated, we should not have it manually generated so we should either follow the same path as IJ Tekton (generate them from Go code) or get them from upstream

**Comment by lstocchi** [comment]
> * how the schema(s) has/have been generated, we should not have it manually generated so we should either follow the same path as IJ Tekton (generate them from Go code) or get them from upstream

I took it from vscode. I remember Joshua worked upstream to create it so the service schema should be ok but he only has this one in vscode so if we want to generate everything with Go code we would be safe for next resources

**Comment by lstocchi** [comment]
@jeffmaury refactored based on your comment. WDYT about the errors?

**Comment by lstocchi** [comment]
This patch works with https://github.com/redhat-developer/intellij-common/pull/47

**Review by evidolob — APPROVED** [review]

**Review by jeffmaury — APPROVED** [review]
LGTM.
Remarks:
- Button in the dialog is 'Add', shouldn't be 'Create' ?
- When the service is created, I did not get the tree refreshed, maybe it's because of my wrong Minikube VM (because it claims my route is invalid (issue with docker image retrieval)
- Noticed you added schemas, we must also add the patch that makes the JetBrains Kubernetes plugin not report errors on Ultimate

**Pull request merged**